blocks/log.py

```diff
@@ -1,274 +1,117 @@
-        self._check_time(time)
-        return TrainingLogRow(self, time)
 
-    @property
-    def current_row(self):
-        return self[self.status.iterations_done]
+class TrainingStatus(Mapping):
+    def __init__(self, exclude=None):
+        self._status = {'iterations_done': 0, 'epochs_done': 0}
+        self.exclude = [] if exclude is None else exclude
 
-    @property
-    def previous_row(self):
-        return self[self.status.iterations_done - 1]
+    def __getitem__(self, key):
+        return self._status[key]
 
-    @property
-    def last_epoch_row(self):
-        return self[self.status._epoch_ends[-1]]
+    def __setitem__(self, key, value):
+        self._status[key] = value
 
-    @abstractmethod
     def __iter__(self):
-        """Returns an iterator over time-key-value triples of the log."""
-        pass
+        return (key for key in self._status if key not in self.exclude)
 
-    @abstractmethod
-    def get_row_iterator(self, time):
-        """Returns an iterator over key-value pairs of a row of the log."""
-        pass
+    def __len__(self):
+        return len(self._status)
 
-    def _check_time(self, time):
-        if not isinstance(time, int) or time < 0:
-            raise ValueError("time must be a positive integer")
 
-    def to_dataframe(self):
-        """Convert a log into a :class:`.DataFrame`."""
-        if not PANDAS_AVAILABLE:
-            raise ImportError("The pandas library is not found. You can"
-                              " install it with pip.")
-        return self._to_dataframe()
-
-    def _to_dataframe(self):
-        raise NotImplementedError()
+class MongoTrainingLog(Mapping):
+    def __init__(self):
+        self.client = MongoClient()
+        self.db = self.client.blocks_log
+        self.entries = self.db.entries
 
+    def __getitem__(self, key):
+        return MongoEntry(self, key)
 
-class TrainingStatus(AbstractTrainingStatus):
-    """A simple training status."""
     def __iter__(self):
-        for attr, value in self.__dict__.items():
-            if not attr.startswith("__"):
-                yield attr, value
-
-
-class TrainingLog(AbstractTrainingLog):
-    """A simple training log storing information in main memory."""
-    def __init__(self):
-        self._storage = defaultdict(dict)
-        self._default_values = {}
-        self._status = TrainingStatus()
+        return map(methodcaller('pop', '_id'),
+                   self.entries.find(projection=['_id'],
+                                     sort=[('_id', ASCENDING)]))
 
-    def get_default_value(self, key):
-        return self._default_values.get(key)
+    def __len__(self):
+        return self.entries.count()
 
-    def set_default_value(self, key, value):
-        self._default_values[key] = value
 
-    def _add_record(self, time, key, value):
-        self._storage[time][key] = value
+class MongoEntry(Mapping):
+    def __init__(self, log, key):
+        self.log = log
+        self.key = key
```

Comment:
It's confusing that you call both time and name keys, when `self.key` is time and `key` is name.